Add minigames/bosses/cities/dungeons to Discord plugin

PandahRS opened the pull request:
Expanding the discord plugin, partially based on issue #2854 
I was thinking of doing boss lairs first, but based Adam suggested to make it for an area instead.

Looking into making a toggle for several types of area's. 
All of them should be an **opt-in** in the configuration, as to not intrude into the privacy too much.

Types such as:
- [x] Bosses
- [x] Cities
- [x]  Dungeons
- More suggestions are always welcome. 

The following example is for Taverly Dungeon:
![](https://i.imgur.com/HuwUr9y.png)

To do list to indicate progress.

- [x] Implement Area types.
- [x] Implement Area type toggles.

While the list of accepted area's could always be expanded upon, it would be nice to have a big list to start off with, with any missing area's being defined by issues.

Previews:
![screenshot 2018-07-04 15 38 09](https://user-images.githubusercontent.com/5115805/42280275-93568310-7fa0-11e8-976f-ff055b0096fa.png)
![screenshot 2018-07-04 15 34 42](https://user-images.githubusercontent.com/5115805/42280277-93907930-7fa0-11e8-956b-1f58a1d0eee8.png)
![screenshot 2018-07-04 15 33 46](https://user-images.githubusercontent.com/5115805/42280278-93b36224-7fa0-11e8-9e0c-7c62973485a4.png)

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -24,36 +24,16 @@
-import static net.runelite.api.Skill.STRENGTH;
-import static net.runelite.api.Skill.THIEVING;
-import static net.runelite.api.Skill.WOODCUTTING;
+import static net.runelite.api.Skill.*;
```
I think this is against the style guide

LeviSchuck commented:
Some areas, such as the cooking guild, are much smaller than a single map region.

PandahRS commented:
That is true, which means either i'd have to make the whole region count as the guild or not at all. thoughts?

LeviSchuck commented:
Maybe multiple methods of finding a named area.
One that is map zone based, maybe one that is map-chunk based. 

On that note, zone 10293 is the fishing guild.

LeviSchuck commented:
The GE is in zone 12598

PandahRS commented:
Extended the configuration in the last commit to include toggles and a list of cities. 
![](https://i.imgur.com/NeSwTCL.png)

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -185,4 +264,15 @@ public static DiscordGameEventType fromSkill(final Skill skill)
+
+	public static  DiscordGameEventType fromLocation(final int regionId)
+	{
+		Optional<DiscordGameEventType>optional = Arrays.stream(DiscordGameEventType.values()).filter(d -> d.regionIds.contains(regionId)).findAny();
```
Dat spacing and line length

LeviSchuck replied:
Also, I'm pretty sure Adam won't like this.
Try iterating through the values, return what matches your filter, or return null.

I think the line count will be about the same for this too.

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +145,33 @@ public void checkForValidStatus()
+			return;
+		}
+
+		final DiscordGameEventType discordGameEventType = DiscordGameEventType.fromLocation(client.getLocalPlayer().getWorldLocation().getRegionID());
```
Line is long and doesn't fit within github review page :1st_place_medal:

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +145,33 @@ public void checkForValidStatus()
+
+		switch (discordGameEventType.getDiscordAreaType())
+		{
+			case BOSSES: if (!config.showBossActivity()) return;
```
The cases will leak across if the config is not on for that area type.

You could do something like..
```
boolean enabled = true;
switch(discordGameEventType.getDiscordAreaType()){
case BOSSES: enabled = config.showBossActivity(); break;
case CITIES: enabled = config.showCityActivity(); break;
...
}
if (!enabled) {
return;
}
```

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +145,33 @@ public void checkForValidStatus()
+		}
+		lastEvent = discordGameEventType;
+		discordState.triggerEvent(discordGameEventType, config.actionDelay());
+
```
Empty line spacing is very weird in many of your files.

LeviSchuck commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -143,6 +206,22 @@ private static String imageKeyOf(final String what)
+		{
+			case BOSSES:
+				return "Fighting " + areaName;
+			case CITIES:
```
I don't really consider being in a city "Exploring"

LeviSchuck commented:
Hmm, what about the wilderness?

PandahRS commented:
Was thinking of the wilderness, but the issue is 

- Several bosses can be moved out of their original spawns, which would not help to indicate where they would be.
- Not sure what cities and dungeons exist in the wilderness, besides the  Wilderness GWD.

forsco commented:
You wouldn't want to 'advertise' being in the wilderness. At least I know I wouldn't.

PandahRS commented:
That too :)

PandahRS commented:
A **thicc** list of dungeons have been added, of which only one has a region filled.
I will start working through them from friday night onwards, but if anyone has some of them laying around feel free to let me know 😄 

![](https://i.imgur.com/ETe60xK.png)

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -62,31 +42,195 @@
+	DUNGEON_DORGESHUUN_MINES("Dorgeshuun Mines", DiscordAreaType.DUNGEONS, ImmutableList.of()),
+	DUNGEON_DRAYNOR_SEWERS("Draynor Sewers", DiscordAreaType.DUNGEONS, ImmutableList.of()),
+	DUNGEON_DWARVEN_MINES("Dwarven Mines", DiscordAreaType.DUNGEONS, ImmutableList.of()),
+	DUNGEON_EAGLES_PEAK("Eagles' Peak Dungeon", DiscordAreaType.DUNGEONS, ImmutableList.of()),
```
Add new constructor override without last param, all these empty immutable lists are weird

PandahRS replied:
These should all be filled with RegionIds though. Any which will not contain a RegionId I will remove.

deathbeam replied:
Ahh

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -123,6 +271,16 @@
+	}
+
+
+
```
Change these 3 new lines to 1

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordAreaType.java:
```diff
@@ -0,0 +1,10 @@
+package net.runelite.client.plugins.discord;
```
Add license header

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +145,42 @@ public void checkForValidStatus()
 		}
 	}
 
+	private void checkForAreaUpdate()
```
This should combine with skilling so skilling will not override it every time

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -124,7 +127,7 @@ public void onXpChanged(ExperienceChanged event)
 		final DiscordGameEventType discordGameEventType = DiscordGameEventType.fromSkill(event.getSkill());
 
-		if (discordGameEventType != null)
+		if (discordGameEventType != null && config.showSkillingActivity())
```
just early return at the top of the method

PandahRS commented:
Besides fixing things for @deathbeam , the trained skill and locations are now split.
![](https://i.imgur.com/SwAQ1tI.png) 

An `onConfigChanged` event is also implemented to clean up the presence message whenever a configuration changes, as not to display past information when it shouldn't be.

Whenever all comments are completed and it's 'acceptable', I will expand on the current list of dungeons and minigames within this PR.

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordAreaType.java:
```diff
@@ -0,0 +1,33 @@
+/*
+ * Copyright (c) 2018, Tomas Slusny <slusnucky@gmail.com>
```
You might want to use your name :P

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -24,77 +24,231 @@
+	BOSS_VORKATH("Vorkath", DiscordAreaType.BOSSES, ImmutableList.of(9023)),
+	BOSS_ZULRAH("Zulrah", DiscordAreaType.BOSSES, ImmutableList.of(9007)),
+
+
```
Remove this empty line

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -109,6 +115,20 @@ protected void shutDown() throws Exception
+		if (event.getGroup().equalsIgnoreCase("discord"))
+		{
+			lastEvent = DiscordGameEventType.IN_GAME;
+			lastEvent.setImageKey("default");
```
Why you need to set imageKey here?

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -165,4 +217,38 @@ else if (client.getGameState() == GameState.LOGGED_IN && (force || !loggedIn))
+	{
+		if (lastSkillEvent != null && config.showSkillingActivity())
+		{
+			lastEvent.setState(lastSkillEvent.getState());
```
Dont mutate basically constants (enum == constants). Store the lastEvent.getState and lastEvent.getImageKey separately.

PandahRS commented:
Haven't been progressing on listing down dungeons and such much. Been busy with irl stuff and recomping at Solak on RS3 while my second monitor is broken as well 😞. 

Would like to work further on it but I have no clue whenever that will be. So an idea would be to either finish the PR with bosses (excluding raids, kinda stuck on that one)/cities or wait until I have the time to map dungeons again. Thoughts?

permedd commented:
I do like this but i also don’t because if someone does black chins at wildy it will show and thats not a great idea.

- Maybe add what world they are in?


-Maybe let it work for safezone **only** and not wilderness?

PandahRS commented:
Finally completed my big boi list of dungeons 👍

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -24,77 +24,219 @@
 	private Function<DiscordGameEventType, Boolean> isChanged = (l) -> true;
 	private int priority = 0;
 
+
```
remove this new line

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -24,77 +24,219 @@
 
+
+	private DiscordAreaType discordAreaType;
+	private List<Integer> regionIds = new ArrayList<>();
```
probably dont initialize this when its optional and do early returns in case its null

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -123,11 +142,14 @@ public void onXpChanged(ExperienceChanged event)
 		final DiscordGameEventType discordGameEventType = DiscordGameEventType.fromSkill(event.getSkill());
-
-		if (discordGameEventType != null)
+		if (discordGameEventType != null && config.showSkillingActivity())
```
add empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +164,40 @@ public void checkForValidStatus()
+		{
+			return;
+		}
+		int playerRegionID = client.getLocalPlayer().getWorldLocation().getRegionID();
```
add empty line above, make this final

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +164,40 @@ public void checkForValidStatus()
+			return;
+		}
+		int playerRegionID = client.getLocalPlayer().getWorldLocation().getRegionID();
+		if (playerRegionID == 0)
```
add empty line above

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -142,6 +164,40 @@ public void checkForValidStatus()
+		}
+
+		lastRegionId = playerRegionID;
+		DiscordGameEventType discordGameEventType = DiscordGameEventType.fromLocation(playerRegionID);
```
mark this final

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -153,16 +209,65 @@ public void flushDiscordStatus()
+			discordState.triggerEvent(lastEvent, config.actionDelay());
+		}
+
+
```
remove this 2 empty lines

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordPlugin.java:
```diff
@@ -153,16 +209,65 @@ public void flushDiscordStatus()
+			discordState.setLastImageKey("default");
+		}
+
+		if (lastAreaEvent != null && showArea(lastAreaEvent))
```
no need to null check lastAreaEvent as showArea has null check inside

NathenSample commented:
Would be nice to see the different areas this defines as `DiscordGameEventType` moved to a generic constant and then referenced here, so that in the futures other are able to use them without re-inventing the wheel.

PandahRS commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordAreaType.java:
```diff
@@ -0,0 +1,32 @@
+ */
+package net.runelite.client.plugins.discord;
+
+public enum DiscordAreaType
```
Or am I supposed to do something like
````java
public class DiscordArea
{
	private enum DiscordAreaType
	{
		BOSSES,
		CITIES,
		DUNGEONS,
	}
}

Abextm replied:
Just
```
enum DiscordAreaType
{
	BOSSES,
	CITIES,
	DUNGEONS,
}
```

PandahRS commented:
> Would be nice to see the different areas this defines as DiscordGameEventType moved to a generic constant class and then referenced here, so that in the futures other are able to use them without re-inventing the wheel.

I agree, should it be put in a constant class as something discord specific or do we consider the locations  as something more general (something like `GameLocation`, `GameArea` `RegionArea`

deathbeam commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordGameEventType.java:
```diff
@@ -24,103 +25,262 @@
 	}
 
-	DiscordGameEventType(Skill skill, Function<DiscordGameEventType, Boolean> isChanged)
+	DiscordGameEventType(String areaName, DiscordAreaType areaType, Integer... regionIds)
```
this is intentional so Arrays.asList works without it being too painful

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordState.java:
```diff
@@ -24,109 +24,129 @@
 
-		lastAction = Instant.now();
-		final DiscordGameEventType newEvent = lastQueue.get(lastQueue.size() - 1);
+		events.sort((a, b) -> ComparisonChain.start()
```
`events.sort(Comparator.comparing(v->v.getType().getPriority()).thenComparing(EventWithTime::getUpdated))`?

Abextm commented on runelite-client/src/main/java/net/runelite/client/plugins/discord/DiscordState.java:
```diff
@@ -24,109 +24,129 @@
-		final Duration actionTimeout = Duration.ofMinutes(timeout);
-
-		return Instant.now().isAfter(lastAction.plus(actionTimeout));
+			final Duration actionTimeout = Duration.ofMinutes(config.actionTimeout());
```
This should be hoisted outside of the loop

deathbeam commented:
It will show only latest skill trained, so probably str, I think and it will also show that you are in NMZ. For slayer it will show you are training slayer and w/e location you are at.

nlruk commented:
Can you display the duel arena with winning/losing streaks?
